olive/engine/cloud_cache_helper.py

```diff
@@ -0,0 +1,191 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import io
+import json
+import logging
+import tempfile
+from copy import deepcopy
+from pathlib import Path
+from typing import Any, Dict
+
+from olive.common.config_utils import ConfigBase
+from olive.common.utils import get_credentials, hash_dict
+from olive.model.config.model_config import ModelConfig
+
+logger = logging.getLogger(__name__)
+
+
+class CloudCacheConfig(ConfigBase):
+    enable_cloud_cache: bool = True
```

Comment:
I would recommend splitting this flag into two - `enable_read` and `enable_write`. This gives better control to user when they want to use the cache but not contribute to it and vice-versa.

This can especially be useful for developers who are doing trial and error runs where they know the output might be not useful.

Reply:
Do you mean `upload_to_cloud` flag?

Reply:
Yes, something along the lines of `enable_upload_to_cloud` and `enable_download_from_cloud`
or `enable_cloud_cache_read` and `enable_cloud_cache_write`

Reply:
`enable_download_from_cloud` would be always true when you enabling cloud cache. And `enable_upload_to_cloud ` is same as `upload_to_cloud ` flag we already have.

Reply:
Use cases -
* If I iterating on a workflow, I could use `enable_download_from_cloud` but disable `enable_upload_to_cloud` to avoid polluting the cache with bad results.
*  If I have already had a successful workflow post iterations with `enable_upload_to_cloud` disabled, I could run the workflow again with `enable_upload_to_cloud` enabled to push the results to cache so that recurring runs can use the cache. The workflow wouldn't do anything because it will use results from previous run but the results should now be uploaded and cached.

Reply:
For your use cases:

- User can enable cloud model cache with true flag in Engine, but set `upload_to_cloud` to false.
- This use case is not being considered when I implement this. Remote cache is mainly focused on downlaoding but not uploading, which means the cache workflow is like: run a pass -> check local cache -> if yes, continue workflow, if no, check cloud cache -> if neither, run pass locally. So we don't have a use that it exist in local cache but we want to update the cloud cache with local cache.

Reply:
Let's reconsider this if/when we have a use case.